python/src/worker/response.py

```diff
@@ -12,53 +13,46 @@
 
 
 class Response:
-    def __init__(self, request, results):
+    def __init__(self, request, result):
         self.request = request
-        self.results = results
+        self.result = result
 
-        self.error = False
-        for result in self.results:
-            self.error = self.error or result.error
-
-        if self.error:
-            self.cacheable = False
-        else:
-            self.cacheable = True
-            for result in self.results:
-                self.cacheable = self.cacheable and result.cacheable
+        self.error = result.error
+        self.cacheable = (not result.error) and result.cacheable
 
         self.s3_bucket = config.RESULTS_BUCKET
 
-    def _construct_response_msg(self, brief=False):
+    def _construct_data_for_upload(self):
+        info("Dumping json data into a string")
+        json_body = json.dumps(self.result.data)
+
+        info("Starting compression before upload to s3")
+        gzipped_body = gzip.compress(json_body.encode('utf-8'))
```

Comment:
you should be able to avoid having to dump the json into a string, and instead compressing it directly doing something similar to:
```
import io
json_body = io.BytesIO()
with gzip.open(json_body, 'wt', encoding="utf-i") as zipfile:
    json.dump(self.result.data, zipfile)
```